File: app/src/main/java/com/stedi/randomimagegenerator/app/model/repository/CachedPresetRepository.java
package com.stedi.randomimagegenerator.app.model.repository;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.SparseArray;

import com.stedi.randomimagegenerator.app.model.data.Preset;
import com.stedi.randomimagegenerator.app.other.Utils;

import java.util.List;

public class CachedPresetRepository implements PresetRepository {
    private final SparseArray<Preset> cache = new SparseArray<>();

    private final PresetRepository target;

    private boolean isActual;

    public CachedPresetRepository(@NonNull PresetRepository target) {
        this.target = target;
    }

    @Override
    public synchronized void save(@NonNull Preset preset) throws Exception {
        target.save(preset);
        cache.put(preset.getId(), preset);
    }

    @Override
    public synchronized void remove(int id) throws Exception {
        target.remove(id);
        cache.remove(id);
    }

    @Override
    @Nullable
    public synchronized Preset get(int id) throws Exception {
        if (cache.indexOfKey(id) >= 0) {
            return cache.get(id);
        } else {
            Preset preset = target.get(id);
            cache.put(id, preset);
            return preset;
        }
    }

    @Override
    @NonNull
    public synchronized List<Preset> getAll() throws Exception {
        if (isActual) {
            return Utils.sparseArrayToList(cache);
        } else {
            List<Preset> result = target.getAll();
            for (Preset preset : result)
                cache.put(preset.getId(), preset);
            isActual = true;
            return result;
        }
    }
}

File: app/src/main/java/com/stedi/randomimagegenerator/app/model/repository/CachedPresetRepository.kt
package com.stedi.randomimagegenerator.app.model.repository

import android.support.annotation.VisibleForTesting
import android.util.SparseArray
import com.stedi.randomimagegenerator.app.model.data.Preset
import com.stedi.randomimagegenerator.app.other.toList

class CachedPresetRepository(private val target: PresetRepository) : PresetRepository {
    private val cache = SparseArray<Preset>()

    private var isActual = false

    @Throws(Exception::class)
    @Synchronized
    override fun save(preset: Preset) {
        target.save(preset)
        cache.put(preset.id, preset)
    }

    @Throws(Exception::class)
    @Synchronized
    override fun remove(id: Int) {
        target.remove(id)
        cache.remove(id)
    }

    @Throws(Exception::class)
    @Synchronized
    override fun get(id: Int): Preset? {
        if (cache.indexOfKey(id) >= 0) {
            return cache.get(id)
        } else {
            val preset = target.get(id)
            if (preset != null) {
                cache.put(id, preset)
            }
            return preset
        }
    }

    @Throws(Exception::class)
    @Synchronized
    override fun getAll(): List<Preset> {
        if (isActual) {
            return cache.toList()
        } else {
            val result = target.getAll()
            for (preset in result) {
                cache.put(preset.id, preset)
            }
            isActual = true
            return result
        }
    }

    @VisibleForTesting
    @Synchronized
    fun getCache() = cache
}